main.go

```diff
@@ -23,76 +23,162 @@ func readFileLines(path string) ([]string, error) {
+				return
+			} else {
+				phraseResults <- PhraseReadResults{parsedPhrase, nil}
+			}
+		}
+	}()
+
+	return phraseResults
+}
+
+type PhraseWav struct {
+	phrase            *datatypes.Phrase
+	phraseWavFilePath string
+}
+
+type PhraseWavResults struct {
+	phraseWav *PhraseWav
+	err       error
+}
+
+/**
+ * Management function for spawning go routines to generate WAV files and manage the channel to do so
+ *
+ * The channel is safe to read until completion
+ */
+func generateWavFilesForPhrases(tpv *datatypes.TaskProcessVariables, phraseResults chan PhraseReadResults) chan PhraseWavResults {
+	phrasesGenerated := make(chan PhraseWavResults)
+
+	go func() {
+		defer close(phrasesGenerated)
+
+		var wg sync.WaitGroup
+
+		for phraseResult := range phraseResults {
+			if phraseResult.err != nil {
+				phrasesGenerated <- PhraseWavResults{nil, phraseResult.err}
+				return
+			}
+
+			phrase := phraseResult.phrase
+
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				generateWavFileForPhrase(tpv, phrase, phrasesGenerated)
+			}()
+		}
+
+		wg.Wait()
+	}()
+
+	return phrasesGenerated
+}
+
+func generateWavFileForPhrase(tpv *datatypes.TaskProcessVariables, phrase *datatypes.Phrase, phrasesGenerated chan PhraseWavResults) {
+	err := (*tpv.Generator).GenerateAudioFile(tpv.Parameters, phrase.PhraseText, tpv.GetPhraseFilePath(phrase.PhraseIndex))
+
+	if err != nil {
+		phrasesGenerated <- PhraseWavResults{nil, err}
+		return
 	}
 
-	phraseResults <- PhraseReadResults{parsedPhrases, nil}
+	phrasesGenerated <- PhraseWavResults{&PhraseWav{phrase, tpv.GetPhraseFilePath(phrase.PhraseIndex)}, nil}
 }
 
+/**
+ * Process task pipeline
+ *
+ * Goes through initializing and managing each step:
+ * - Convert input audio to WAV
+ * - Parse provided phrases
+ * - Generate WAV files from parsed phrases
+ * - Prepare for MFCC
+ * - Prepare for DTW
+ */
 func processTask(results chan string, task *datatypes.Task, generator *datatypes.AudioGenerator, tempDir string) {
 	tpv := datatypes.NewTaskProcessVariables(task, generator, tempDir)
 	defer func() {
 		results <- tpv.GetFinalLogs()
 	}()
 
 	if len(task.Description) > 0 {
-		tpv.Println("")
-		tpv.Println("*** ", tpv.Task.Description, " ***")
-		tpv.Println("")
+		fmt.Println("")
+		fmt.Println("*** ", tpv.Task.Description, " ***")
+		fmt.Println("")
 	}
-	tpv.Println("Audio   : ", tpv.Task.AudioFilename)
-	tpv.Println("Phrase  : ", tpv.Task.PhraseFilename)
-	tpv.Println("Output  : ", tpv.Task.OutputFilename)
-	tpv.Println("Parameters : ", tpv.Parameters)
+	fmt.Println("Audio   : ", tpv.Task.AudioFilename)
+	fmt.Println("Phrase  : ", tpv.Task.PhraseFilename)
+	fmt.Println("Output  : ", tpv.Task.OutputFilename)
+	fmt.Println("Parameters : ", tpv.Parameters)
 
 	wavs := make(chan string)
 	go convertWav(wavs, tpv)
 
-	phrases := make(chan PhraseReadResults)
-	go readPhrasesFromFile(tpv.Task.PhraseFilename, phrases)
+	parsedPhrases := readPhrasesFromFile(tpv.Task.PhraseFilename)
```

Comment:
No goroutines?

Couldn't you take the output channel from the readPhrasesFromFile as an input channel into the generateWavFilesFromPhrases?

Reply:
The readPhrasesFromFile creates the channel and spawns a goroutine internally, so that the management of the goroutine can happen inside of the function, e.g. on lines 60-61